{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Author:\n",
    "    Diana Y. Lee, Luque Lab, SDSU\n",
    "    dlee@sdsu.edu\n",
    "\n",
    "##### Purpose:\n",
    "    Base random forest model for predicting the capsid architecture (as measured by the \n",
    "    T-number) of a tailed phage from the MCP sequence\n",
    "\n",
    "##### Requires: \n",
    "\n",
    "    phage_functions.ipynb  :  Functions for calculating T based on genome size\n",
    "    \n",
    "    MCP2T_RF_state.db  :  Trained random forest model database\n",
    "\n",
    "    data\\PHAGE_DATA.csv : phage data with indexes, genome size, and translations. Must include the following columns:\n",
    "        'genome_length'\n",
    "        'MCP_len'\n",
    "        'Virus_ID'\n",
    "        'MCP_Sequence'\n",
    "\n",
    "    data\\isoelectric_points.csv : Isoelectric point data for the phage table. Must include the following columns:\n",
    "        'Virus_ID'\n",
    "        'IPC'\n",
    "\n",
    "    NOTE: Current jupyter notebook includes commands for renaming the necessary columns to match this convention.\n",
    "   \n",
    "##### Creates:\n",
    "    results\\phageResults.csv  :  results of the random forest prediction\n",
    "    results\\phageResults_db.db  :  database state after the prediction\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "np.random.seed(42)\n",
    "import csv\n",
    "import time\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ML imports\n",
    "from sklearn.metrics import mean_squared_error\n",
    "from sklearn import metrics\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.model_selection import cross_val_score\n",
    "from sklearn.model_selection import cross_val_predict\n",
    "from sklearn.base import BaseEstimator\n",
    "from sklearn.metrics import precision_score, recall_score, f1_score\n",
    "from sklearn.metrics import precision_recall_curve\n",
    "from sklearn.metrics import roc_curve\n",
    "from sklearn.metrics import roc_auc_score\n",
    "from sklearn.metrics import confusion_matrix\n",
    "from sklearn.metrics import mean_squared_error\n",
    "from sklearn.metrics import accuracy_score\n",
    "from sklearn.model_selection import cross_val_score, GridSearchCV\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "from sklearn.model_selection import cross_val_score\n",
    "from sklearn import metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqUtils import GC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from ipynb.fs.full.phage_functions import tNearest\n",
    "from ipynb.fs.full.phage_functions import tNearestValid\n",
    "from ipynb.fs.full.phage_functions import tModel\n",
    "from ipynb.fs.full.phage_functions import tNum\n",
    "from ipynb.fs.full.phage_functions import tList\n",
    "from ipynb.fs.full.phage_functions import tDictAll"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# custom function to count amino acids\n",
    "# amino acids are hardcoded to avoid broken dependencies, since they do not change\n",
    "def createFreq(acidSeq, normF=None):\n",
    "    normF = normF or 0\n",
    "    if (normF > 1):\n",
    "        print(\"Valid tTypes are 0 (raw) or 1 (normalized by sample). Defaults to 0.\")\n",
    "        return\n",
    "    AA = []\n",
    "    aaList = np.asarray(['A', 'R', 'N', 'D', 'C', 'Q', 'E', 'G', 'H', 'I', 'L', 'K', 'M', 'F', 'P', 'S', 'T', 'W', 'Y', 'V'])\n",
    "    aaLen=len(aaList)\n",
    "    n = len(acidSeq)\n",
    "    for i in range(n):\n",
    "        for acid in (aaList):\n",
    "            trans1=str(acidSeq[i])\n",
    "            a = trans1.count(str(acid))\n",
    "            AA.append(a)\n",
    "    rFreq = np.asarray(AA).reshape((n, aaLen))\n",
    "    if (normF == 0):\n",
    "#        print(\"Success! Created an nx20 array, where n is the length of the list provided:\",n)\n",
    "#        print(\"Columns are frequency totals for each amino acid:\",aaList)\n",
    "        return rFreq\n",
    "    if (normF == 1):\n",
    "        nFreq = copy.copy(rFreq).astype(float)\n",
    "        fff3 = copy.copy(rFreq).astype(float)\n",
    "        nf = rFreq.shape[1]\n",
    "        for i in range(nf):\n",
    "            nFreq[:,i] = fff3[:,i]/fff3.sum(axis=1)\n",
    "#        print(\"Success! Created an nx20 array, where n is the length of the list provided:\",n)\n",
    "#        print(\"Columns are frequency percentages for each amino acid:\",aaList)\n",
    "        return nFreq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# custom function to create dataset with only sequence length, frequency, and isoelectric point\n",
    "# requires a dataframe with fields \"Virus_Name\", MCP_Sequence\",\"IPC\", and \"MCP_len\"\n",
    "\n",
    "def createDataset3(dF):\n",
    "    nn=dF.shape[0]\n",
    "    freq = createFreq(dF[\"MCP_Sequence\"], 1)\n",
    "    AAT = []\n",
    "    for i in range(nn):\n",
    "        AAT.append(dF.iloc[i][\"Virus_ID\"])\n",
    "        AAT.append(dF.iloc[i][\"IPC\"])\n",
    "        AAT.append(dF.iloc[i][\"MCP_len\"])\n",
    "        for j in range(20):\n",
    "            AAT.append(freq[i][j])\n",
    "        AAT.append(dF.iloc[i][\"T_nearest_errMar_code\"])\n",
    "    AAT = np.reshape(np.ravel(AAT), (nn, 24));\n",
    "    AAT = np.asarray(AAT)\n",
    "\n",
    "    \n",
    "#    print(\"Success! Created an nx24 array, where n is the length of the list provided:\",n)\n",
    "#    print(\"Column 0: Virus_Name\")\n",
    "#    print(\"Column 1: Isoelectric Point\")\n",
    "#    print(\"Column 2: length of MCP sequence\")\n",
    "#    print(\"Columns 3-22 are frequency percentages for each amino acid\")\n",
    "#    print(\"Column 23: Target T\")\n",
    "    return AAT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "## load kernel state\n",
    "## rerun imports cell as well\n",
    "import dill\n",
    "dill.load_session('MCP2T_RF_state.db')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BG_72112_Abt2graduatex2\n",
      "567\n"
     ]
    }
   ],
   "source": [
    "# import a single fasta for an MCP (protein format)\n",
    "for record in SeqIO.parse(\"data/test_MCP.faa\", \"fasta\"):\n",
    "    print(record.description)\n",
    "    print(len(record.seq))\n",
    "    test_ID = record.description\n",
    "    test_MCP = (str)(record.seq)\n",
    "    test_MCP_len = len(record.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add the iso\n",
    "test_isopt = 4.23\n",
    "# add the genome length if you have it (in bps)\n",
    "test_genomeLen = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calc the T numbers based on the genomeLen\n",
    "t1 = round(tNum(test_genomeLen/1000,0),4)\n",
    "t2 = tNum(test_genomeLen/1000,1)\n",
    "t3 = tNum(test_genomeLen/1000,2,errMar)\n",
    "t4 = tdict2[tNum(test_genomeLen/1000,2,errMar)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Virus_ID</th>\n",
       "      <th>MCP_Sequence</th>\n",
       "      <th>MCP_len</th>\n",
       "      <th>IPC</th>\n",
       "      <th>genome_length</th>\n",
       "      <th>T_raw</th>\n",
       "      <th>T_nearest</th>\n",
       "      <th>T_nearest_errMar</th>\n",
       "      <th>T_nearest_errMar_code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BG_72112_Abt2graduatex2</td>\n",
       "      <td>MADMYELPDDVTALTDDDLDANLAAAVRSFSTVSRTTVVTPQTLPN...</td>\n",
       "      <td>567</td>\n",
       "      <td>4.23</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0036</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Virus_ID                                       MCP_Sequence  \\\n",
       "0  BG_72112_Abt2graduatex2  MADMYELPDDVTALTDDDLDANLAAAVRSFSTVSRTTVVTPQTLPN...   \n",
       "\n",
       "   MCP_len   IPC  genome_length   T_raw  T_nearest  T_nearest_errMar  \\\n",
       "0      567  4.23              1  0.0036        1.0               1.0   \n",
       "\n",
       "   T_nearest_errMar_code  \n",
       "0                      0  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create a dataframe for the single phage\n",
    "d = {'Virus_ID': [test_ID], 'MCP_Sequence': [test_MCP], 'MCP_len': [test_MCP_len], 'IPC': [test_isopt],'genome_length': [test_genomeLen],\"T_raw\" : t1, \"T_nearest\" : t2, \"T_nearest_errMar\" : t3, \"T_nearest_errMar_code\" : t4}\n",
    "oneDF = pd.DataFrame(data=d)\n",
    "\n",
    "oneDF[\"T_raw\"] = oneDF[\"T_raw\"].astype('float64')\n",
    "oneDF[\"T_nearest\"] = oneDF[\"T_nearest\"].astype('float64')\n",
    "oneDF[\"T_nearest_errMar\"] = oneDF[\"T_nearest_errMar\"].astype('float64')\n",
    "oneDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create the dataset for the random forest\n",
    "x_OnePhage = createDataset3(oneDF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# assign the features and labels\n",
    "x_OneActual = (x_OnePhage[0:1,1:23]).astype(float)\n",
    "y_OneActual = (x_OnePhage[0:1,23]).astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MCP2T-RF predicts this phage to be T= 13.0\n"
     ]
    }
   ],
   "source": [
    "# get the random forest prediction\n",
    "y_onePhagePredCode = rfBest_clf.predict(x_OneActual)\n",
    "y_onePhagePred = tdict2rev[y_onePhagePredCode[0]]\n",
    "\n",
    "print(\"MCP2T-RF predicts this phage to be T=\", y_onePhagePred)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# predict T numbers for a whole list of MCPs\n",
    "# import phage data\n",
    "phageData = pd.read_csv(\"data\\PHAGE_DATA.csv\")\n",
    "# remove records from the dataframe if the ID is nan\n",
    "for i in range(len(phageData[\"ID\"])):\n",
    "    if(np.isnan(phageData[\"ID\"][i])):\n",
    "        phageData = phageData.drop(index=i)\n",
    "# get a count\n",
    "n = len(phageData[\"ID\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>GBK_ID</th>\n",
       "      <th>CDS</th>\n",
       "      <th>DEFINITION</th>\n",
       "      <th>LOCUS_COMPLETE_GENOME</th>\n",
       "      <th>COMPLETE_GENOME_BP</th>\n",
       "      <th>ORGANISM</th>\n",
       "      <th>NCBI_GENPEPT_PROTEIN_ID</th>\n",
       "      <th>PROTEIN_PRODUCT</th>\n",
       "      <th>GENE_LOCUS</th>\n",
       "      <th>PROTEIN_BP</th>\n",
       "      <th>TRANSLATION</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1262521.3</td>\n",
       "      <td>bp:11683..12636</td>\n",
       "      <td>Leuconostoc phage phiLNTR3, complete genome.</td>\n",
       "      <td>NC_024378_1_28015</td>\n",
       "      <td>28015</td>\n",
       "      <td>Viruses; dsDNA viruses, no RNA stage; Caudovir...</td>\n",
       "      <td>YP_009044222.1</td>\n",
       "      <td>putative major capsid protein</td>\n",
       "      <td>HL53_gp15</td>\n",
       "      <td>317</td>\n",
       "      <td>MGIEFLSTSKAVELYAKLALETQGNTETFSRKWKDIVSERSEQAIT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1273740.3</td>\n",
       "      <td>bp:9604..10494</td>\n",
       "      <td>Bacillus phage Curly, complete genome.</td>\n",
       "      <td>NC_020479_1_49425</td>\n",
       "      <td>49425</td>\n",
       "      <td>Viruses; dsDNA viruses, no RNA stage; Caudovir...</td>\n",
       "      <td>YP_007517560.1</td>\n",
       "      <td>major capsid protein</td>\n",
       "      <td>CURLY_16</td>\n",
       "      <td>296</td>\n",
       "      <td>MADIVLGQHPLLKKVFLDRRIKDFTASGFVADQLFTNISVDALAIK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>1282994.3</td>\n",
       "      <td>bp:13216..14253</td>\n",
       "      <td>Burkholderia phage ST79, complete genome.</td>\n",
       "      <td>NC_021343_1_35430</td>\n",
       "      <td>35430</td>\n",
       "      <td>Viruses; dsDNA viruses, no RNA stage; Caudovir...</td>\n",
       "      <td>YP_008060494.1</td>\n",
       "      <td>major capsid protein</td>\n",
       "      <td>M190_gp20</td>\n",
       "      <td>345</td>\n",
       "      <td>MNPITRRALTRYMDNIAKLNGVASVAEKFAVAPSVQQTLEKRIQES...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>633135.2</td>\n",
       "      <td>bp:4469..5662</td>\n",
       "      <td>Streptococcus phage Abc2</td>\n",
       "      <td>NC_013645_1_34882</td>\n",
       "      <td>34882</td>\n",
       "      <td>Viruses; dsDNA viruses, no RNA stage; Caudovir...</td>\n",
       "      <td>YP_003347415.1</td>\n",
       "      <td>Phage major capsid protein # ACLAME 20</td>\n",
       "      <td>SP-Abc2_gp06</td>\n",
       "      <td>397</td>\n",
       "      <td>MKTSNELHDLWVAQGDKVENLNEKLNVAMLDDSVTAEELQKIKNER...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>1168563.3</td>\n",
       "      <td>bp:5642..6655</td>\n",
       "      <td>Stenotrophomonas phage Smp131, complete genome.</td>\n",
       "      <td>NC_023588_1_33525</td>\n",
       "      <td>33525</td>\n",
       "      <td>Viruses; dsDNA viruses, no RNA stage; Caudovir...</td>\n",
       "      <td>YP_009008364.1</td>\n",
       "      <td>phage major capsid protein precursor</td>\n",
       "      <td>CH36_gp09</td>\n",
       "      <td>337</td>\n",
       "      <td>MRTKTRRLFEGYTQQVATLNNVSGVANTFSVEPTVQQSLEARMQES...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID     GBK_ID              CDS  \\\n",
       "0   0  1262521.3  bp:11683..12636   \n",
       "1   1  1273740.3   bp:9604..10494   \n",
       "2   2  1282994.3  bp:13216..14253   \n",
       "3   3   633135.2    bp:4469..5662   \n",
       "4   5  1168563.3    bp:5642..6655   \n",
       "\n",
       "                                        DEFINITION LOCUS_COMPLETE_GENOME  \\\n",
       "0     Leuconostoc phage phiLNTR3, complete genome.     NC_024378_1_28015   \n",
       "1           Bacillus phage Curly, complete genome.     NC_020479_1_49425   \n",
       "2        Burkholderia phage ST79, complete genome.     NC_021343_1_35430   \n",
       "3                         Streptococcus phage Abc2     NC_013645_1_34882   \n",
       "4  Stenotrophomonas phage Smp131, complete genome.     NC_023588_1_33525   \n",
       "\n",
       "   COMPLETE_GENOME_BP                                           ORGANISM  \\\n",
       "0               28015  Viruses; dsDNA viruses, no RNA stage; Caudovir...   \n",
       "1               49425  Viruses; dsDNA viruses, no RNA stage; Caudovir...   \n",
       "2               35430  Viruses; dsDNA viruses, no RNA stage; Caudovir...   \n",
       "3               34882  Viruses; dsDNA viruses, no RNA stage; Caudovir...   \n",
       "4               33525  Viruses; dsDNA viruses, no RNA stage; Caudovir...   \n",
       "\n",
       "  NCBI_GENPEPT_PROTEIN_ID                         PROTEIN_PRODUCT  \\\n",
       "0          YP_009044222.1           putative major capsid protein   \n",
       "1          YP_007517560.1                    major capsid protein   \n",
       "2          YP_008060494.1                    major capsid protein   \n",
       "3          YP_003347415.1  Phage major capsid protein # ACLAME 20   \n",
       "4          YP_009008364.1    phage major capsid protein precursor   \n",
       "\n",
       "     GENE_LOCUS  PROTEIN_BP                                        TRANSLATION  \n",
       "0     HL53_gp15         317  MGIEFLSTSKAVELYAKLALETQGNTETFSRKWKDIVSERSEQAIT...  \n",
       "1      CURLY_16         296  MADIVLGQHPLLKKVFLDRRIKDFTASGFVADQLFTNISVDALAIK...  \n",
       "2     M190_gp20         345  MNPITRRALTRYMDNIAKLNGVASVAEKFAVAPSVQQTLEKRIQES...  \n",
       "3  SP-Abc2_gp06         397  MKTSNELHDLWVAQGDKVENLNEKLNVAMLDDSVTAEELQKIKNER...  \n",
       "4     CH36_gp09         337  MRTKTRRLFEGYTQQVATLNNVSGVANTFSVEPTVQQSLEARMQES...  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# exmaine your phage data and see if any changes are necessary to the required column names \n",
    "# ('genome_length', 'MCP_len', 'Virus_ID', 'MCP_Sequence')\n",
    "\n",
    "phageData[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# change any necessary column names using this command, with the arguments formatted as {\"original_column_name\" : \"New_name\"}\n",
    "phageData = phageData.rename(columns={\"COMPLETE_GENOME_BP\": 'genome_length',\"PROTEIN_BP\": 'MCP_len',\"ID\": 'Virus_ID','TRANSLATION':'MCP_Sequence'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import the isoelectric points\n",
    "isoelectricPT = pd.read_csv(\"data\\isoelectric_points4.csv\")\n",
    "isoelectricPT = isoelectricPT.rename(columns={\"header\": \"ID\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sequence</th>\n",
       "      <th>molecular_weight</th>\n",
       "      <th>average_pI</th>\n",
       "      <th>IPC_protein</th>\n",
       "      <th>IPC_peptide</th>\n",
       "      <th>Toseland</th>\n",
       "      <th>Thurlkill</th>\n",
       "      <th>Nozaki_Tanford</th>\n",
       "      <th>Dawson</th>\n",
       "      <th>...</th>\n",
       "      <th>Patrickios</th>\n",
       "      <th>Rodwell</th>\n",
       "      <th>Sillero</th>\n",
       "      <th>Solomon</th>\n",
       "      <th>Lehninger</th>\n",
       "      <th>Wikipedia</th>\n",
       "      <th>ProMoST</th>\n",
       "      <th>pH_5.5_charge</th>\n",
       "      <th>pH_7.4_charge</th>\n",
       "      <th>pH_8.0_charge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>MGIEFLSTSKAVELYAKLALETQGNTETFSRKWKDIVSERSEQAIT...</td>\n",
       "      <td>34756.64584</td>\n",
       "      <td>5.032</td>\n",
       "      <td>5.107</td>\n",
       "      <td>5.074</td>\n",
       "      <td>4.995</td>\n",
       "      <td>4.988</td>\n",
       "      <td>5.177</td>\n",
       "      <td>5.066</td>\n",
       "      <td>...</td>\n",
       "      <td>4.439</td>\n",
       "      <td>4.973</td>\n",
       "      <td>5.246</td>\n",
       "      <td>5.066</td>\n",
       "      <td>5.022</td>\n",
       "      <td>4.935</td>\n",
       "      <td>5.220</td>\n",
       "      <td>-2.9</td>\n",
       "      <td>-6.4</td>\n",
       "      <td>-7.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>MADIVLGQHPLLKKVFLDRRIKDFTASGFVADQLFTNISVDALAIK...</td>\n",
       "      <td>33091.30464</td>\n",
       "      <td>4.748</td>\n",
       "      <td>4.835</td>\n",
       "      <td>4.801</td>\n",
       "      <td>4.687</td>\n",
       "      <td>4.692</td>\n",
       "      <td>4.906</td>\n",
       "      <td>4.798</td>\n",
       "      <td>...</td>\n",
       "      <td>4.081</td>\n",
       "      <td>4.682</td>\n",
       "      <td>4.961</td>\n",
       "      <td>4.796</td>\n",
       "      <td>4.752</td>\n",
       "      <td>4.685</td>\n",
       "      <td>4.955</td>\n",
       "      <td>-6.3</td>\n",
       "      <td>-10.4</td>\n",
       "      <td>-11.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>MNPITRRALTRYMDNIAKLNGVASVAEKFAVAPSVQQTLEKRIQES...</td>\n",
       "      <td>38088.20534</td>\n",
       "      <td>5.548</td>\n",
       "      <td>5.606</td>\n",
       "      <td>5.618</td>\n",
       "      <td>5.528</td>\n",
       "      <td>5.580</td>\n",
       "      <td>5.737</td>\n",
       "      <td>5.612</td>\n",
       "      <td>...</td>\n",
       "      <td>4.427</td>\n",
       "      <td>5.522</td>\n",
       "      <td>5.809</td>\n",
       "      <td>5.611</td>\n",
       "      <td>5.571</td>\n",
       "      <td>5.504</td>\n",
       "      <td>5.742</td>\n",
       "      <td>0.5</td>\n",
       "      <td>-3.6</td>\n",
       "      <td>-5.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>MKTSNELHDLWVAQGDKVENLNEKLNVAMLDDSVTAEELQKIKNER...</td>\n",
       "      <td>44206.00734</td>\n",
       "      <td>4.937</td>\n",
       "      <td>5.004</td>\n",
       "      <td>4.973</td>\n",
       "      <td>4.868</td>\n",
       "      <td>4.868</td>\n",
       "      <td>5.077</td>\n",
       "      <td>4.969</td>\n",
       "      <td>...</td>\n",
       "      <td>4.522</td>\n",
       "      <td>4.858</td>\n",
       "      <td>5.136</td>\n",
       "      <td>4.968</td>\n",
       "      <td>4.923</td>\n",
       "      <td>4.855</td>\n",
       "      <td>5.125</td>\n",
       "      <td>-5.6</td>\n",
       "      <td>-10.7</td>\n",
       "      <td>-13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>MRTKTRRLFEGYTQQVATLNNVSGVANTFSVEPTVQQSLEARMQES...</td>\n",
       "      <td>37689.78144</td>\n",
       "      <td>8.850</td>\n",
       "      <td>8.374</td>\n",
       "      <td>9.293</td>\n",
       "      <td>8.841</td>\n",
       "      <td>9.033</td>\n",
       "      <td>9.088</td>\n",
       "      <td>9.149</td>\n",
       "      <td>...</td>\n",
       "      <td>4.900</td>\n",
       "      <td>9.251</td>\n",
       "      <td>9.205</td>\n",
       "      <td>9.306</td>\n",
       "      <td>9.268</td>\n",
       "      <td>9.310</td>\n",
       "      <td>8.738</td>\n",
       "      <td>9.7</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID                                           sequence  molecular_weight  \\\n",
       "0   0  MGIEFLSTSKAVELYAKLALETQGNTETFSRKWKDIVSERSEQAIT...       34756.64584   \n",
       "1   1  MADIVLGQHPLLKKVFLDRRIKDFTASGFVADQLFTNISVDALAIK...       33091.30464   \n",
       "2   2  MNPITRRALTRYMDNIAKLNGVASVAEKFAVAPSVQQTLEKRIQES...       38088.20534   \n",
       "3   3  MKTSNELHDLWVAQGDKVENLNEKLNVAMLDDSVTAEELQKIKNER...       44206.00734   \n",
       "4   5  MRTKTRRLFEGYTQQVATLNNVSGVANTFSVEPTVQQSLEARMQES...       37689.78144   \n",
       "\n",
       "   average_pI  IPC_protein  IPC_peptide  Toseland  Thurlkill  Nozaki_Tanford  \\\n",
       "0       5.032        5.107        5.074     4.995      4.988           5.177   \n",
       "1       4.748        4.835        4.801     4.687      4.692           4.906   \n",
       "2       5.548        5.606        5.618     5.528      5.580           5.737   \n",
       "3       4.937        5.004        4.973     4.868      4.868           5.077   \n",
       "4       8.850        8.374        9.293     8.841      9.033           9.088   \n",
       "\n",
       "   Dawson  ...  Patrickios  Rodwell  Sillero  Solomon  Lehninger  Wikipedia  \\\n",
       "0   5.066  ...       4.439    4.973    5.246    5.066      5.022      4.935   \n",
       "1   4.798  ...       4.081    4.682    4.961    4.796      4.752      4.685   \n",
       "2   5.612  ...       4.427    5.522    5.809    5.611      5.571      5.504   \n",
       "3   4.969  ...       4.522    4.858    5.136    4.968      4.923      4.855   \n",
       "4   9.149  ...       4.900    9.251    9.205    9.306      9.268      9.310   \n",
       "\n",
       "   ProMoST  pH_5.5_charge  pH_7.4_charge  pH_8.0_charge  \n",
       "0    5.220           -2.9           -6.4           -7.9  \n",
       "1    4.955           -6.3          -10.4          -11.7  \n",
       "2    5.742            0.5           -3.6           -5.8  \n",
       "3    5.125           -5.6          -10.7          -13.0  \n",
       "4    8.738            9.7            3.0            1.5  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# exmaine your phageisoelectric data and see if any changes are necessary to the required column names \n",
    "# ('Virus_ID', 'IPC')\n",
    "isoelectricPT[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# change any necessary column names using this command, with the arguments formatted as {\"original_column_name\" : \"New_name\"}\n",
    "isoelectricPT = isoelectricPT.rename(columns={\"IPC_protein\": 'IPC',\"ID\": 'Virus_ID'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add the isoelectric points to the dataframe\n",
    "phageData = phageData.merge(isoelectricPT[[\"Virus_ID\",\"IPC\"]], on=\"Virus_ID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Virus_ID</th>\n",
       "      <th>T_raw</th>\n",
       "      <th>T_nearest</th>\n",
       "      <th>T_nearest_errMar</th>\n",
       "      <th>T_nearest_errMar_code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.2455</td>\n",
       "      <td>5.33</td>\n",
       "      <td>5.33</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>7.8613</td>\n",
       "      <td>7.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2.0</td>\n",
       "      <td>6.2010</td>\n",
       "      <td>7.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3.0</td>\n",
       "      <td>6.1325</td>\n",
       "      <td>5.33</td>\n",
       "      <td>0.00</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5.0</td>\n",
       "      <td>5.9615</td>\n",
       "      <td>5.33</td>\n",
       "      <td>0.00</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Virus_ID   T_raw  T_nearest  T_nearest_errMar  T_nearest_errMar_code\n",
       "0       0.0  5.2455       5.33              5.33                      4\n",
       "1       1.0  7.8613       7.00              0.00                     47\n",
       "2       2.0  6.2010       7.00              0.00                     47\n",
       "3       3.0  6.1325       5.33              0.00                     47\n",
       "4       5.0  5.9615       5.33              0.00                     47"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# calculate T numbers\n",
    "ny = phageData.shape[0]\n",
    "Y_T = []\n",
    "\n",
    "for i in range(ny):\n",
    "    Y_T.append(phageData.iloc[i][\"Virus_ID\"])\n",
    "    Y_T.append(round(tNum(phageData.iloc[i][\"genome_length\"]/1000,0),4))\n",
    "    Y_T.append(tNum(phageData.iloc[i][\"genome_length\"]/1000,1))\n",
    "    Y_T.append(tNum(phageData.iloc[i][\"genome_length\"]/1000,2,errMar))\n",
    "    Y_T.append(tdict2[tNum(phageData.iloc[i][\"genome_length\"]/1000,2,errMar)])\n",
    "    \n",
    "Y = np.asarray(Y_T)\n",
    "Y = np.reshape(np.ravel(Y), (ny, 5));\n",
    "Y = np.asarray(Y)\n",
    "\n",
    "df_T = pd.DataFrame(Y)\n",
    "df_T = df_T.rename(columns={0: 'Virus_ID', 1: 'T_raw', 2: 'T_nearest', 3: 'T_nearest_errMar', 4: 'T_nearest_errMar_code'})\n",
    "\n",
    "df_T[\"T_raw\"] = df_T[\"T_raw\"].astype('float64')\n",
    "df_T[\"T_nearest\"] = df_T[\"T_nearest\"].astype('float64')\n",
    "df_T[\"T_nearest_errMar\"] = df_T[\"T_nearest_errMar\"].astype('float64')\n",
    "df_T[\"T_nearest_errMar_code\"] = df_T[\"T_nearest_errMar_code\"].astype('int64')\n",
    "\n",
    "df_T [0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Virus_ID</th>\n",
       "      <th>GBK_ID</th>\n",
       "      <th>CDS</th>\n",
       "      <th>DEFINITION</th>\n",
       "      <th>LOCUS_COMPLETE_GENOME</th>\n",
       "      <th>genome_length</th>\n",
       "      <th>ORGANISM</th>\n",
       "      <th>NCBI_GENPEPT_PROTEIN_ID</th>\n",
       "      <th>PROTEIN_PRODUCT</th>\n",
       "      <th>GENE_LOCUS</th>\n",
       "      <th>MCP_len</th>\n",
       "      <th>MCP_Sequence</th>\n",
       "      <th>IPC</th>\n",
       "      <th>T_raw</th>\n",
       "      <th>T_nearest</th>\n",
       "      <th>T_nearest_errMar</th>\n",
       "      <th>T_nearest_errMar_code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1262521.3</td>\n",
       "      <td>bp:11683..12636</td>\n",
       "      <td>Leuconostoc phage phiLNTR3, complete genome.</td>\n",
       "      <td>NC_024378_1_28015</td>\n",
       "      <td>28015</td>\n",
       "      <td>Viruses; dsDNA viruses, no RNA stage; Caudovir...</td>\n",
       "      <td>YP_009044222.1</td>\n",
       "      <td>putative major capsid protein</td>\n",
       "      <td>HL53_gp15</td>\n",
       "      <td>317</td>\n",
       "      <td>MGIEFLSTSKAVELYAKLALETQGNTETFSRKWKDIVSERSEQAIT...</td>\n",
       "      <td>5.107</td>\n",
       "      <td>5.2455</td>\n",
       "      <td>5.33</td>\n",
       "      <td>5.33</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1273740.3</td>\n",
       "      <td>bp:9604..10494</td>\n",
       "      <td>Bacillus phage Curly, complete genome.</td>\n",
       "      <td>NC_020479_1_49425</td>\n",
       "      <td>49425</td>\n",
       "      <td>Viruses; dsDNA viruses, no RNA stage; Caudovir...</td>\n",
       "      <td>YP_007517560.1</td>\n",
       "      <td>major capsid protein</td>\n",
       "      <td>CURLY_16</td>\n",
       "      <td>296</td>\n",
       "      <td>MADIVLGQHPLLKKVFLDRRIKDFTASGFVADQLFTNISVDALAIK...</td>\n",
       "      <td>4.835</td>\n",
       "      <td>7.8613</td>\n",
       "      <td>7.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>1282994.3</td>\n",
       "      <td>bp:13216..14253</td>\n",
       "      <td>Burkholderia phage ST79, complete genome.</td>\n",
       "      <td>NC_021343_1_35430</td>\n",
       "      <td>35430</td>\n",
       "      <td>Viruses; dsDNA viruses, no RNA stage; Caudovir...</td>\n",
       "      <td>YP_008060494.1</td>\n",
       "      <td>major capsid protein</td>\n",
       "      <td>M190_gp20</td>\n",
       "      <td>345</td>\n",
       "      <td>MNPITRRALTRYMDNIAKLNGVASVAEKFAVAPSVQQTLEKRIQES...</td>\n",
       "      <td>5.606</td>\n",
       "      <td>6.2010</td>\n",
       "      <td>7.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>633135.2</td>\n",
       "      <td>bp:4469..5662</td>\n",
       "      <td>Streptococcus phage Abc2</td>\n",
       "      <td>NC_013645_1_34882</td>\n",
       "      <td>34882</td>\n",
       "      <td>Viruses; dsDNA viruses, no RNA stage; Caudovir...</td>\n",
       "      <td>YP_003347415.1</td>\n",
       "      <td>Phage major capsid protein # ACLAME 20</td>\n",
       "      <td>SP-Abc2_gp06</td>\n",
       "      <td>397</td>\n",
       "      <td>MKTSNELHDLWVAQGDKVENLNEKLNVAMLDDSVTAEELQKIKNER...</td>\n",
       "      <td>5.004</td>\n",
       "      <td>6.1325</td>\n",
       "      <td>5.33</td>\n",
       "      <td>0.00</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>1168563.3</td>\n",
       "      <td>bp:5642..6655</td>\n",
       "      <td>Stenotrophomonas phage Smp131, complete genome.</td>\n",
       "      <td>NC_023588_1_33525</td>\n",
       "      <td>33525</td>\n",
       "      <td>Viruses; dsDNA viruses, no RNA stage; Caudovir...</td>\n",
       "      <td>YP_009008364.1</td>\n",
       "      <td>phage major capsid protein precursor</td>\n",
       "      <td>CH36_gp09</td>\n",
       "      <td>337</td>\n",
       "      <td>MRTKTRRLFEGYTQQVATLNNVSGVANTFSVEPTVQQSLEARMQES...</td>\n",
       "      <td>8.374</td>\n",
       "      <td>5.9615</td>\n",
       "      <td>5.33</td>\n",
       "      <td>0.00</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Virus_ID     GBK_ID              CDS  \\\n",
       "0         0  1262521.3  bp:11683..12636   \n",
       "1         1  1273740.3   bp:9604..10494   \n",
       "2         2  1282994.3  bp:13216..14253   \n",
       "3         3   633135.2    bp:4469..5662   \n",
       "4         5  1168563.3    bp:5642..6655   \n",
       "\n",
       "                                        DEFINITION LOCUS_COMPLETE_GENOME  \\\n",
       "0     Leuconostoc phage phiLNTR3, complete genome.     NC_024378_1_28015   \n",
       "1           Bacillus phage Curly, complete genome.     NC_020479_1_49425   \n",
       "2        Burkholderia phage ST79, complete genome.     NC_021343_1_35430   \n",
       "3                         Streptococcus phage Abc2     NC_013645_1_34882   \n",
       "4  Stenotrophomonas phage Smp131, complete genome.     NC_023588_1_33525   \n",
       "\n",
       "   genome_length                                           ORGANISM  \\\n",
       "0          28015  Viruses; dsDNA viruses, no RNA stage; Caudovir...   \n",
       "1          49425  Viruses; dsDNA viruses, no RNA stage; Caudovir...   \n",
       "2          35430  Viruses; dsDNA viruses, no RNA stage; Caudovir...   \n",
       "3          34882  Viruses; dsDNA viruses, no RNA stage; Caudovir...   \n",
       "4          33525  Viruses; dsDNA viruses, no RNA stage; Caudovir...   \n",
       "\n",
       "  NCBI_GENPEPT_PROTEIN_ID                         PROTEIN_PRODUCT  \\\n",
       "0          YP_009044222.1           putative major capsid protein   \n",
       "1          YP_007517560.1                    major capsid protein   \n",
       "2          YP_008060494.1                    major capsid protein   \n",
       "3          YP_003347415.1  Phage major capsid protein # ACLAME 20   \n",
       "4          YP_009008364.1    phage major capsid protein precursor   \n",
       "\n",
       "     GENE_LOCUS  MCP_len                                       MCP_Sequence  \\\n",
       "0     HL53_gp15      317  MGIEFLSTSKAVELYAKLALETQGNTETFSRKWKDIVSERSEQAIT...   \n",
       "1      CURLY_16      296  MADIVLGQHPLLKKVFLDRRIKDFTASGFVADQLFTNISVDALAIK...   \n",
       "2     M190_gp20      345  MNPITRRALTRYMDNIAKLNGVASVAEKFAVAPSVQQTLEKRIQES...   \n",
       "3  SP-Abc2_gp06      397  MKTSNELHDLWVAQGDKVENLNEKLNVAMLDDSVTAEELQKIKNER...   \n",
       "4     CH36_gp09      337  MRTKTRRLFEGYTQQVATLNNVSGVANTFSVEPTVQQSLEARMQES...   \n",
       "\n",
       "     IPC   T_raw  T_nearest  T_nearest_errMar  T_nearest_errMar_code  \n",
       "0  5.107  5.2455       5.33              5.33                      4  \n",
       "1  4.835  7.8613       7.00              0.00                     47  \n",
       "2  5.606  6.2010       7.00              0.00                     47  \n",
       "3  5.004  6.1325       5.33              0.00                     47  \n",
       "4  8.374  5.9615       5.33              0.00                     47  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add T predictions to the phage data\n",
    "currentDataset = phageData.merge(df_T, how='left', on='Virus_ID')\n",
    "currentDataset[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create the dataset for the random forest\n",
    "x_Phage = createDataset3(currentDataset)\n",
    "\n",
    "# assign the features and labels\n",
    "x_actual = (x_Phage[0:ny,1:23]).astype(float)\n",
    "y_actual = (x_Phage[0:ny,23]).astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RandomForestClassifier(class_weight='balanced', max_depth=20, max_features=6,\n",
      "                       min_samples_leaf=3, min_samples_split=6,\n",
      "                       n_estimators=750, random_state=42)\n",
      "RandomForestClassifier Accuracy:  0.9173419773095624\n"
     ]
    }
   ],
   "source": [
    "# rfBest_clf=RandomForestClassifier(random_state = 42, max_features=6,n_estimators=750, max_depth=20, \n",
    "#                       min_samples_split=6, min_samples_leaf=3,bootstrap = True, class_weight='balanced')rfBest_clf.fit(x_train, y_train)\n",
    "# is already trained on the data\n",
    "yPred = rfBest_clf.predict(x_actual)\n",
    "print(rfBest_clf)\n",
    "print(rfBest_clf.__class__.__name__, \"Accuracy: \", accuracy_score(y_actual, yPred))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# predict T-numbers for your data\n",
    "Y_Predictions = []\n",
    "for i in range(ny):\n",
    "    Y_Predictions.append(x_Phage[0:ny,0][i].astype(int))\n",
    "    Y_Predictions.append(tdict2rev[yPred[i]])\n",
    "    Y_Predictions.append(yPred[i])\n",
    "    \n",
    "Y_Predictions = np.asarray(np.reshape(np.ravel(Y_Predictions), (ny, 3)))\n",
    "Y_Predictions = pd.DataFrame(Y_Predictions)\n",
    "Y_Predictions = Y_Predictions.rename(columns={0: 'Virus_ID',1:'T_Pred',2:'T_Pred_Code'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Virus_ID</th>\n",
       "      <th>GBK_ID</th>\n",
       "      <th>CDS</th>\n",
       "      <th>DEFINITION</th>\n",
       "      <th>LOCUS_COMPLETE_GENOME</th>\n",
       "      <th>genome_length</th>\n",
       "      <th>ORGANISM</th>\n",
       "      <th>NCBI_GENPEPT_PROTEIN_ID</th>\n",
       "      <th>PROTEIN_PRODUCT</th>\n",
       "      <th>GENE_LOCUS</th>\n",
       "      <th>MCP_len</th>\n",
       "      <th>MCP_Sequence</th>\n",
       "      <th>IPC</th>\n",
       "      <th>T_raw</th>\n",
       "      <th>T_nearest</th>\n",
       "      <th>T_nearest_errMar</th>\n",
       "      <th>T_nearest_errMar_code</th>\n",
       "      <th>T_Pred</th>\n",
       "      <th>T_Pred_Code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1262521.3</td>\n",
       "      <td>bp:11683..12636</td>\n",
       "      <td>Leuconostoc phage phiLNTR3, complete genome.</td>\n",
       "      <td>NC_024378_1_28015</td>\n",
       "      <td>28015</td>\n",
       "      <td>Viruses; dsDNA viruses, no RNA stage; Caudovir...</td>\n",
       "      <td>YP_009044222.1</td>\n",
       "      <td>putative major capsid protein</td>\n",
       "      <td>HL53_gp15</td>\n",
       "      <td>317</td>\n",
       "      <td>MGIEFLSTSKAVELYAKLALETQGNTETFSRKWKDIVSERSEQAIT...</td>\n",
       "      <td>5.107</td>\n",
       "      <td>5.2455</td>\n",
       "      <td>5.33</td>\n",
       "      <td>5.33</td>\n",
       "      <td>4</td>\n",
       "      <td>5.33</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1273740.3</td>\n",
       "      <td>bp:9604..10494</td>\n",
       "      <td>Bacillus phage Curly, complete genome.</td>\n",
       "      <td>NC_020479_1_49425</td>\n",
       "      <td>49425</td>\n",
       "      <td>Viruses; dsDNA viruses, no RNA stage; Caudovir...</td>\n",
       "      <td>YP_007517560.1</td>\n",
       "      <td>major capsid protein</td>\n",
       "      <td>CURLY_16</td>\n",
       "      <td>296</td>\n",
       "      <td>MADIVLGQHPLLKKVFLDRRIKDFTASGFVADQLFTNISVDALAIK...</td>\n",
       "      <td>4.835</td>\n",
       "      <td>7.8613</td>\n",
       "      <td>7.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>47</td>\n",
       "      <td>0.00</td>\n",
       "      <td>47.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>1282994.3</td>\n",
       "      <td>bp:13216..14253</td>\n",
       "      <td>Burkholderia phage ST79, complete genome.</td>\n",
       "      <td>NC_021343_1_35430</td>\n",
       "      <td>35430</td>\n",
       "      <td>Viruses; dsDNA viruses, no RNA stage; Caudovir...</td>\n",
       "      <td>YP_008060494.1</td>\n",
       "      <td>major capsid protein</td>\n",
       "      <td>M190_gp20</td>\n",
       "      <td>345</td>\n",
       "      <td>MNPITRRALTRYMDNIAKLNGVASVAEKFAVAPSVQQTLEKRIQES...</td>\n",
       "      <td>5.606</td>\n",
       "      <td>6.2010</td>\n",
       "      <td>7.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>47</td>\n",
       "      <td>7.00</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>633135.2</td>\n",
       "      <td>bp:4469..5662</td>\n",
       "      <td>Streptococcus phage Abc2</td>\n",
       "      <td>NC_013645_1_34882</td>\n",
       "      <td>34882</td>\n",
       "      <td>Viruses; dsDNA viruses, no RNA stage; Caudovir...</td>\n",
       "      <td>YP_003347415.1</td>\n",
       "      <td>Phage major capsid protein # ACLAME 20</td>\n",
       "      <td>SP-Abc2_gp06</td>\n",
       "      <td>397</td>\n",
       "      <td>MKTSNELHDLWVAQGDKVENLNEKLNVAMLDDSVTAEELQKIKNER...</td>\n",
       "      <td>5.004</td>\n",
       "      <td>6.1325</td>\n",
       "      <td>5.33</td>\n",
       "      <td>0.00</td>\n",
       "      <td>47</td>\n",
       "      <td>0.00</td>\n",
       "      <td>47.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>1168563.3</td>\n",
       "      <td>bp:5642..6655</td>\n",
       "      <td>Stenotrophomonas phage Smp131, complete genome.</td>\n",
       "      <td>NC_023588_1_33525</td>\n",
       "      <td>33525</td>\n",
       "      <td>Viruses; dsDNA viruses, no RNA stage; Caudovir...</td>\n",
       "      <td>YP_009008364.1</td>\n",
       "      <td>phage major capsid protein precursor</td>\n",
       "      <td>CH36_gp09</td>\n",
       "      <td>337</td>\n",
       "      <td>MRTKTRRLFEGYTQQVATLNNVSGVANTFSVEPTVQQSLEARMQES...</td>\n",
       "      <td>8.374</td>\n",
       "      <td>5.9615</td>\n",
       "      <td>5.33</td>\n",
       "      <td>0.00</td>\n",
       "      <td>47</td>\n",
       "      <td>0.00</td>\n",
       "      <td>47.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Virus_ID     GBK_ID              CDS  \\\n",
       "0         0  1262521.3  bp:11683..12636   \n",
       "1         1  1273740.3   bp:9604..10494   \n",
       "2         2  1282994.3  bp:13216..14253   \n",
       "3         3   633135.2    bp:4469..5662   \n",
       "4         5  1168563.3    bp:5642..6655   \n",
       "\n",
       "                                        DEFINITION LOCUS_COMPLETE_GENOME  \\\n",
       "0     Leuconostoc phage phiLNTR3, complete genome.     NC_024378_1_28015   \n",
       "1           Bacillus phage Curly, complete genome.     NC_020479_1_49425   \n",
       "2        Burkholderia phage ST79, complete genome.     NC_021343_1_35430   \n",
       "3                         Streptococcus phage Abc2     NC_013645_1_34882   \n",
       "4  Stenotrophomonas phage Smp131, complete genome.     NC_023588_1_33525   \n",
       "\n",
       "   genome_length                                           ORGANISM  \\\n",
       "0          28015  Viruses; dsDNA viruses, no RNA stage; Caudovir...   \n",
       "1          49425  Viruses; dsDNA viruses, no RNA stage; Caudovir...   \n",
       "2          35430  Viruses; dsDNA viruses, no RNA stage; Caudovir...   \n",
       "3          34882  Viruses; dsDNA viruses, no RNA stage; Caudovir...   \n",
       "4          33525  Viruses; dsDNA viruses, no RNA stage; Caudovir...   \n",
       "\n",
       "  NCBI_GENPEPT_PROTEIN_ID                         PROTEIN_PRODUCT  \\\n",
       "0          YP_009044222.1           putative major capsid protein   \n",
       "1          YP_007517560.1                    major capsid protein   \n",
       "2          YP_008060494.1                    major capsid protein   \n",
       "3          YP_003347415.1  Phage major capsid protein # ACLAME 20   \n",
       "4          YP_009008364.1    phage major capsid protein precursor   \n",
       "\n",
       "     GENE_LOCUS  MCP_len                                       MCP_Sequence  \\\n",
       "0     HL53_gp15      317  MGIEFLSTSKAVELYAKLALETQGNTETFSRKWKDIVSERSEQAIT...   \n",
       "1      CURLY_16      296  MADIVLGQHPLLKKVFLDRRIKDFTASGFVADQLFTNISVDALAIK...   \n",
       "2     M190_gp20      345  MNPITRRALTRYMDNIAKLNGVASVAEKFAVAPSVQQTLEKRIQES...   \n",
       "3  SP-Abc2_gp06      397  MKTSNELHDLWVAQGDKVENLNEKLNVAMLDDSVTAEELQKIKNER...   \n",
       "4     CH36_gp09      337  MRTKTRRLFEGYTQQVATLNNVSGVANTFSVEPTVQQSLEARMQES...   \n",
       "\n",
       "     IPC   T_raw  T_nearest  T_nearest_errMar  T_nearest_errMar_code  T_Pred  \\\n",
       "0  5.107  5.2455       5.33              5.33                      4    5.33   \n",
       "1  4.835  7.8613       7.00              0.00                     47    0.00   \n",
       "2  5.606  6.2010       7.00              0.00                     47    7.00   \n",
       "3  5.004  6.1325       5.33              0.00                     47    0.00   \n",
       "4  8.374  5.9615       5.33              0.00                     47    0.00   \n",
       "\n",
       "   T_Pred_Code  \n",
       "0          4.0  \n",
       "1         47.0  \n",
       "2          5.0  \n",
       "3         47.0  \n",
       "4         47.0  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add the prediction results to the currentDataset\n",
    "phageDataResult = currentDataset.merge(Y_Predictions, how='left', on='Virus_ID')\n",
    "phageDataResult[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# exports results to csv\n",
    "phageDataResult.to_csv(r'results\\phageResultsTest.csv', index=False)\n",
    "# saves kernel state\n",
    "import dill\n",
    "dill.dump_session('results\\phageResultsTest_db.db')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
